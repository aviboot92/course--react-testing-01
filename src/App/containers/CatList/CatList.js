import React from 'react'
import styled from 'styled-components'
import { Pane, toaster, SelectMenu, Button } from 'evergreen-ui'
import getCopy from '../../helpers/getCopy'
import CatsService from '../../services/CatService'
import ProductCard from '../../components/ProductCard/ProductCard'

export default class CatList extends React.Component {
	state = {
		cats: [],
		selectedDescription: undefined,
	}

	async componentDidMount() {
		const cats = await CatsService.getCats()
		this.setState({ cats })
	}

	/**
	 * @return {Array<{Object}>}
	 */
	get filteredCats() {
		if (!this.state.selectedDescription) {
			return this.state.cats
		}

		return this.state.cats.filter(cat => {
			return cat.description.toLowerCase().includes(this.state.selectedDescription)
		})
	}

	/**
	 * @returns {Array<{string}>}
	 */
	get availableDescriptions() {
		return this.state.cats.reduce((uniqDescs, cat) => {
			const descs = cat.description.split(', ')
			for (const desc of descs) {
				const lc = desc.toLowerCase()
				if (!uniqDescs.includes(lc)) {
					uniqDescs.push(lc)
				}
			}

			return uniqDescs
		}, [])
	}

	/**
	 * @returns {Array<{ value: string, label: string }>}
	 */
	get descriptionFilterOptions() {
		return this.availableDescriptions.map(label => ({ label, value: label }))
	}

	/**
	 * @returns {string}
	 */
	get descriptionFilterLabel() {
		return this.state.selectedDescription || 'Select description'
	}

	/**
	 * @param {Object<{ imgSrc: string, name: string }} productObj
	 */
	onAdoptClick = cat => {
		toaster.success(`You have successfully adopted ${cat.name}!`)
		const cats = this.state.cats.filter(c => c.image !== cat.imgSrc)
		this.setState({ cats })
	}

	/**
	 * @param {Object<{ value: any, label: string }>} filterItem
	 */
	handleDescriptionFilter = ({ value }) => {
		this.setState({ selectedDescription: value })
	}

	render() {
		return (
			<Styled>
				<Pane className="filters">
					<SelectMenu
						hasTitle={false}
						hasFilter={false}
						options={this.descriptionFilterOptions}
						selected={this.state.selectedDescription}
						onSelect={this.handleDescriptionFilter}
						closeOnSelect={true}
					>
						<Button>{this.descriptionFilterLabel}</Button>
					</SelectMenu>
				</Pane>
				<Pane className="list">
					{this.filteredCats.map(cat => (
						<Pane margin="10px" key={cat.image}>
							<ProductCard
								name={cat.name}
								age={cat.age}
								imgSrc={cat.image}
								description={cat.description}
								ctaText={getCopy('adoptCtaText')}
								onCtaClick={this.onAdoptClick}
							/>
						</Pane>
					))}
				</Pane>
			</Styled>
		)
	}
}

const Styled = styled(Pane)`
	.filters {
		margin: 10px;
	}
	.list {
		display: grid;
		grid-template-columns: 1fr;
		@media (min-width: 768px) {
			grid-template-columns: repeat(2, 1fr);
		}
	}
`
