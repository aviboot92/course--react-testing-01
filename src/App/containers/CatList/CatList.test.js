import React from 'react'
import { shallow } from 'enzyme'
import { toaster, SelectMenu } from 'evergreen-ui'
import CatList from './CatList'

const mockCats = [
  { name: 'Buba', description: 'Affectionate, Shy', age: 5, image: 'image1' },
  { name: 'Mittens', description: 'Affectionate, Cheeky', age: 9, image: 'image2' }
]

jest.mock('../../services/CatService', () => ({
  getCats: () => mockCats
}))

describe('<CatList />', () => {

  describe('componentDidMount()', () => {
    it('should getCats from CatsService and set to state', async () => {
      const instance = await shallow(<CatList />).instance()
      instance.state.cats = []
      await instance.componentDidMount()
      expect(instance.state.cats).toEqual(mockCats)
    })    
  })  

  describe('get filteredCats()', () => {
    const instance = shallow(<CatList />).instance()
    it('should return state.cats when !state.selectedDescription', () => {
      instance.state.cats = mockCats
      instance.state.selectedDescription = undefined
      expect(instance.filteredCats).toEqual(mockCats)
    })
    it('should return cats filtered by description', () => {
      instance.state.cats = mockCats
      instance.state.selectedDescription = 'cheeky'
      expect(instance.filteredCats).toEqual(mockCats.slice(1))
    })
  })

  describe('get availableDescriptions()', () => {
    const instance = shallow(<CatList />).instance()
    it('should return an empty array when no cats', () => {
      instance.state.cats = []
      expect(instance.availableDescriptions).toEqual([])
    })
    it('should return an array of lowercase unique descriptions', () => {
      instance.state.cats = mockCats
      expect(instance.availableDescriptions).toEqual(['affectionate', 'shy', 'cheeky'])
    })
  })

  describe('get descriptionFilterOptions()', () => {
    const instance = shallow(<CatList />).instance()
    it('should return expected array of filter options', () => {
      jest.spyOn(instance, 'availableDescriptions', 'get').mockReturnValue(['a', 'b'])
      expect(instance.descriptionFilterOptions).toEqual([
        { value: 'a', label: 'a' }, { value: 'b', label: 'b' },
      ])
    })  
  })

  describe('get descriptionFilterLabel()', () => {
    const instance = shallow(<CatList />).instance()
    it('should be state.selectedDescription when truthy', () => {
      instance.state.selectedDescription = 'a'
      expect(instance.descriptionFilterLabel).toBe('a')
    })
    it('should default to "Select description"', () => {
      instance.state.selectedDescription = undefined
      expect(instance.descriptionFilterLabel).toBe('Select description')
    })
  })

  describe('onAdoptClick()', () => {
    const instance = shallow(<CatList />).instance()
    const toasterSuccessSpy = jest.spyOn(toaster, 'success')
    beforeEach(() => {
      toasterSuccessSpy.mockClear()
      instance.state.cats = mockCats
    })
    it('should call toaster.success and filter cat from state', () => {
      instance.onAdoptClick({ name: mockCats[0].name, imgSrc: mockCats[0].image })
      expect(toasterSuccessSpy).toHaveBeenCalledWith(`You have successfully adopted ${mockCats[0].name}!`)
      expect(instance.state.cats).toEqual(mockCats.slice(1))
    })
  })

  describe('handleDescriptionFilter()', () => {
    const instance = shallow(<CatList />).instance()
    it('should set value to state.selectedDescription', () => {
      instance.state.selectedDescription = undefined
      instance.handleDescriptionFilter({ value: 'a' })
      expect(instance.state.selectedDescription).toBe('a')
    })
  })

  describe('render()', () => {
    it('should render as expected', async () => {
      const wrapper = await shallow(<CatList />)
      expect(wrapper).toMatchSnapshot()
    })
    it('should render essential elements', async () => {
      const wrapper = await shallow(<CatList />)
      expect(wrapper.find(SelectMenu).length).toBe(1)
      expect(wrapper.find('ProductCard').length).toBe(mockCats.length)
    })
  })
})
