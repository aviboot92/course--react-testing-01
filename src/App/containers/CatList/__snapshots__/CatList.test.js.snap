// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CatList /> render() should render as expected 1`] = `
<Styled(withTheme(Pane))>
  <withTheme(Pane)
    className="filters"
  >
    <SelectMenu
      closeOnSelect={true}
      filterIcon="search"
      filterPlaceholder="Filter..."
      hasFilter={false}
      hasTitle={false}
      height={248}
      isMultiSelect={false}
      onDeselect={[Function]}
      onSelect={[Function]}
      options={
        Array [
          Object {
            "label": "affectionate",
            "value": "affectionate",
          },
          Object {
            "label": "shy",
            "value": "shy",
          },
          Object {
            "label": "cheeky",
            "value": "cheeky",
          },
        ]
      }
      position="bottom-left"
      width={240}
    >
      <withTheme(Button)>
        Select description
      </withTheme(Button)>
    </SelectMenu>
  </withTheme(Pane)>
  <withTheme(Pane)
    className="list"
  >
    <withTheme(Pane)
      key="image1"
      margin="10px"
    >
      <ProductCard
        age={5}
        ctaText="adopt"
        description="Affectionate, Shy"
        imgSrc="image1"
        name="Buba"
        onCtaClick={[Function]}
      />
    </withTheme(Pane)>
    <withTheme(Pane)
      key="image2"
      margin="10px"
    >
      <ProductCard
        age={9}
        ctaText="adopt"
        description="Affectionate, Cheeky"
        imgSrc="image2"
        name="Mittens"
        onCtaClick={[Function]}
      />
    </withTheme(Pane)>
  </withTheme(Pane)>
</Styled(withTheme(Pane))>
`;
